File: dotnet/experiments/akka-actors/Worker.cs
using Akka.Actor;

namespace AkkaActors;

// ---------------------------------------------------------------------------
// Worker actor: a ReceiveActor holding PRIVATE state — a single int counter.
//
// AKKA-vs-hand-rolled HIGHLIGHT #2 (the runtime IS the actor):
//   In the siblings you had to BUILD the actor runtime yourself: a Channel /
//   MailboxProcessor for the mailbox, a background loop that reads one message at
//   a time (run-to-completion), and a try/catch that turns a thrown exception
//   into an "Exit"/"ChildExit" message for the supervisor. Akka supplies ALL of
//   that. You subclass ReceiveActor, register `Receive<T>` handlers, and get:
//     * a mailbox + single-threaded run-to-completion loop (so `_count` needs no
//       lock — only one message is ever in flight),
//     * automatic crash capture: throwing here is caught by Akka and reported to
//       THIS actor's parent as a failure (no hand-written Exit plumbing),
//     * lifecycle hooks (PreStart/PreRestart/PostRestart/PostStop) so you can see
//       and shape the restart.
// ---------------------------------------------------------------------------
public sealed class Worker : ReceiveActor
{
    // PRIVATE per-actor state. A restart throws this whole object away and builds
    // a fresh Worker from its Props, so `_count` is naturally reset to 0 — exactly
    // like "call the factory again" in the hand-rolled siblings.
    private int _count;

    // The actor's name lives on its path (assigned by the parent at ActorOf time),
    // so we read it back for the trace instead of passing it into the constructor.
    private string Name => Self.Path.Name;

    public Worker()
    {
        // Runs on first start AND on every restart (a brand-new instance is built).
        Trace.Line($"    [{Name}] ctor: new instance, fresh state (count={_count})");

        // One handler per message type. Akka routes by runtime type for us.
        Receive<Increment>(m =>
        {
            _count += m.By;
            Trace.Line($"    [{Name}] Increment(+{m.By}) -> count={_count}");
        });

        Receive<GetCount>(_ =>
        {
            // "ask" side: reply to whoever asked. Akka threads the reply address
            // (Sender) through for us — no reply channel embedded in the message.
            Trace.Line($"    [{Name}] GetCount -> {_count}");
            Sender.Tell(_count);
        });

        Receive<Boom>(_ =>
        {
            // Crash on purpose. We just THROW — Akka catches it and escalates the
            // failure to our parent (the Supervisor), which applies its strategy.
            Trace.Line($"    [{Name}] Boom received -> throwing!");
            throw new InvalidOperationException($"{Name} was poisoned at count={_count}");
        });
    }

    // ---- Lifecycle hooks: make the restart visible in the trace --------------

    protected override void PreStart() =>
        Trace.Line($"    [{Name}] PreStart (now accepting messages)");

    // Called on the FAILED instance, just before it is discarded. `_count` here is
    // still the pre-crash value, proving the about-to-be-lost state.
    protected override void PreRestart(Exception reason, object message)
    {
        Trace.Line($"    [{Name}] PreRestart: discarding state (count was {_count}) due to: {reason.Message}");
        base.PreRestart(reason, message); // default: stop children + call PostStop
    }

    // Called on the FRESH instance, just after Akka rebuilds it. `_count` is 0.
    protected override void PostRestart(Exception reason)
    {
        Trace.Line($"    [{Name}] PostRestart: restarted with fresh state (count={_count})");
        base.PostRestart(reason); // default: call PreStart
    }

    protected override void PostStop() =>
        Trace.Line($"    [{Name}] PostStop (instance torn down)");
}

File: dotnet/experiments/fsharp-actors/Worker.fs
namespace FSharpActors

open System

// -----------------------------------------------------------------------------
// Worker actor: a MailboxProcessor<WorkerMsg> with PRIVATE, IMMUTABLE state.
//
// F#-vs-C# HIGHLIGHT #2 (native mailbox + immutable state threading):
//   C# has no built-in actor; you either pull in Akka.NET / Orleans or hand-roll
//   a `Channel<T>` + a background `while(await reader.ReadAsync())` loop and guard
//   mutable fields. Here the runtime primitive IS the actor: `inbox.Receive()`
//   delivers one message at a time (run-to-completion, no locks), and the private
//   counter is just a parameter `count` threaded through the recursive async
//   loop. State is never mutated in place — each turn produces the next state by
//   calling `loop (newCount)`. Restart = call the factory again => fresh `count`.
// -----------------------------------------------------------------------------

module Worker =

    /// A factory builds a *fresh* worker. Restarting a child is literally
    /// "call the factory again": new MailboxProcessor, new mailbox, counter
    /// back at 0. `supervisor` is threaded in so the child can report its own
    /// crash back as a ChildExit message.
    let create (id: int) (supervisor: MailboxProcessor<SupervisorMsg>) : MailboxProcessor<WorkerMsg> =
        MailboxProcessor.Start(fun inbox ->

            // The recursive loop carries the worker's entire private state
            // (here just `count`) as immutable parameters.
            let rec loop (count: int) =
                async {
                    // Receive blocks (asynchronously) until exactly one message
                    // is available, guaranteeing one-at-a-time processing.
                    let! msg = inbox.Receive()

                    // EXHAUSTIVE match — the compiler checks all WorkerMsg cases
                    // are handled.
                    match msg with
                    | Increment n ->
                        let count' = count + n
                        Log.line (sprintf "    [worker %d] Increment %d  -> count = %d" id n count')
                        return! loop count' // tail-call with the NEW immutable state

                    | GetCount reply ->
                        // "ask" side: answer over the channel embedded in the msg.
                        Log.line (sprintf "    [worker %d] GetCount        -> count = %d" id count)
                        reply.Reply count
                        return! loop count

                    | Poison ->
                        Log.line (sprintf "    [worker %d] Poison received -> throwing!" id)
                        // Raising here escapes the async and is caught below.
                        failwithf "worker %d was poisoned" id
                }

            // Turn a crash INTO A MESSAGE. Wrapping the loop in try/with means an
            // exception in the body does not silently kill the actor: we forward
            // it to the supervisor as ChildExit and let the loop terminate. The
            // supervisor then decides whether to restart (one_for_one).
            async {
                try
                    do! loop 0 // fresh state on every (re)start
                with ex ->
                    supervisor.Post(ChildExit(id, ex))
                // Falling off the end here ends the async => this mailbox stops.
            })
